### PR TITLE
Use analyzers from targeting pack for NetCoreAppCurrent

### LINKED ISSUE #87582: FrameworkList.xml generated in libraries build for live ref pack does not include source generators
The FrameworkList.xml file that's generated for the ref pack that the build outputs to `MicrosoftNetCoreAppRefPackDir` does not include any of our analyzer or generator assemblies. As a result, any tests that depend on this ref pack don't reference ref-pack generators by default.

This is particularly difficult for our trimming tests, which have their project files generated and would need modification to the generation script to support this.

Also, solving this would help make our "out-of-band" projects even closer to a "not in dotnet/runtime" project by removing the requirements for them to have project references to in-box generators.

### PR DESCRIPTION
**Pull request by ViktorHofer**
Fixes that analyzer failures didn't show-up in #74897
Fixes https://github.com/dotnet/runtime/issues/87582
Continuation of https://github.com/dotnet/runtime/pull/75093

Add analyzers to the frameworklist that OOB projects in src/libraries use, and only auto ProjectReference the analyzers in generators.targets when not using the analyzers from the targeting pack.

Also move the generator projects related code into a separate file.

### CODE REVIEW

**Comment by ViktorHofer** [comment]
```
/__w/1/s/src/libraries/System.Runtime.InteropServices/tests/TestAssets/SharedTypes/ComInterfaces/IPointProvider.cs(16,15): error SYSLIB1051: Runtime marshalling must be disabled in this project by applying the 'System.Runtime.CompilerServices.DisableRuntimeMarshallingAttribute' to the assembly to enable marshalling this type. The generated source will not handle marshalling of the return value of method 'GetPoint'. [/__w/1/s/src/libraries/System.Runtime.InteropServices/tests/TestAssets/SharedTypes/SharedTypes.csproj]
/__w/1/s/src/libraries/System.Runtime.InteropServices/tests/TestAssets/SharedTypes/ComInterfaces/IPointProvider.cs(18,29): error SYSLIB1051: Runtime marshalling must be disabled in this project by applying the 'System.Runtime.CompilerServices.DisableRuntimeMarshallingAttribute' to the assembly to enable marshalling this type. The generated source will not handle marshalling of parameter 'point'. [/__w/1/s/src/libraries/System.Runtime.InteropServices/tests/TestAssets/SharedTypes/SharedTypes.csproj]
```

@jkoritzinsky I'm not exactly sure what to do with this error. I assume disabling runtime marshalling for the assembly is the wrong thing to do?

File: eng/generators.targets [hunk 1]
```diff
@@ -13,7 +14,7 @@
     <!-- If the current project is not System.Private.CoreLib, we enable the LibraryImportGenerator source generator
          when the project is a C# source project that:
          - doesn't target the latest TFM or
-         - references everything from the targeting pack (OOB) or
+         - doesn't reference the targeting pack (i.e. when inbox) and
          - references System.Private.CoreLib, or
          - references System.Runtime.InteropServices -->
```
Comment (jkoritzinsky) [inline eng/generators.targets:19]: Let's indent these cases to make it clearer:

```suggestion
             - references System.Private.CoreLib, or
             - references System.Runtime.InteropServices -->
```

File: src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/System.Text.RegularExpressions.Tests.csproj [hunk 2]
```diff
@@ -45,7 +45,7 @@
     <Compile Include="RegexParserTests.netfx.cs" />
     <Compile Include="RegexGeneratorHelper.netfx.cs" />
     <Compile Include="$(CoreLibSharedDir)System\Diagnostics\CodeAnalysis\StringSyntaxAttribute.cs" />
-    <PackageReference Include="System.Text.Json" Version="$(SystemTextJsonVersion)" />
+    <ProjectReference Include="$(LibrariesProjectRoot)System.Text.Json\src\System.Text.Json.csproj" />
```
Comment (ViktorHofer) [inline src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/System.Text.RegularExpressions.Tests.csproj:48]: Unrelated change but I noticed that this test library doesn't use the live System.Text.Json for the netfx build.

File: eng/generators.targets [hunk 3]
```diff
@@ -22,28 +23,26 @@
...
+                                        '@(Reference->AnyHaveMetadataValue('Identity', 'System.Runtime.InteropServices'))' == 'true' or
+                                        '@(ProjectReference->AnyHaveMetadataValue('Identity', '$(CoreLibProject)'))' == 'true'
+                                      )
                                     )
                                   )" />
     <!-- We enable the ComInterfaceGenerator source generator
          when the project is a C# source project that:
-         - References everything from the targeting pack (OOB) or
+         - Doesn't reference the targeting pack (i.e. when inbox) and
          - references System.Runtime.InteropServices -->
```
Comment (jkoritzinsky) [inline eng/generators.targets:37]:
```suggestion
         - References System.Runtime.InteropServices directly and not through the targeting pack (i.e. when inbox) -->
```

**Review by jkoritzinsky — APPROVED** [review]
LGTM!

File: eng/targetingpacks.targets [hunk 4]
```diff
@@ -129,12 +129,14 @@
   <Target Name="UpdateLocalTargetingAndRuntimePack"
           Condition="'$(UseLocalTargetingRuntimePack)' == 'true'"
           AfterTargets="ResolveFrameworkReferences">
+    <Error Text="The shared framework must be built before the local targeting pack can be consumed."
+           Condition="!Exists('$(MicrosoftNetCoreAppRefPackDir)data\FrameworkList.xml')" />
```
Comment (ViktorHofer) [inline eng/targetingpacks.targets:133]: Microsoft.NET.WebAssembly.Webcil.csproj unintentionally built against the LKG even though `UseLocalTargetingRuntimePack` is true. Adding the error makes this more explicit.

**Pull request merged**